Apply client side timeouts

johnbcoughlin opened the pull request:
@ejona86 opening this as a work in progress to get some advice.

A couple of questions:

- It seems hard to satisfy the contract for `ClientStreamListener.close`, that it always be the last method called on the listener, if we have a stream racing with a deadline to call it. One solution is to have the listener impl provide its own idempotence via a CAS or some other mechanism, but that obviously doesn't satisfy the public contract on the interface. It's not clear to me how important that is.
- The first call on a Netty channel takes at least 120 millis, due (I think) to `obtainActiveTransport()`. Reading briefly into that code though, I don't see any obvious expensive asynchronous operations. Thoughts on this? It would be bad to always time out the first call on a channel.

ejona86 commented:
The contract to `close()` is very important, but it's also not as hard to satisfy at it would seem.

Firstly, you should realize that almost all stream processing occurs within the transport, and the transport is single-threaded. In Netty's case, this is because it runs on its own thread. In OkHttp, it acquires a lock. So state checking, in general, is already thread-safe.

It appears that OkHttp does not acquire the lock in these code paths, but instead uses an [atomic operation on a map](https://github.com/grpc/grpc-java/blob/master/okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java#L456). I do find it very difficult to reason about the thread-safety of various calls; it requires way too much knowledge of how the pieces interact to verify. But I also do believe that the current code is relatively bug-free of thread safety issues.

Secondly, we guarantee to serialize all callbacks, thus we can even [have non-volatile state](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/ChannelImpl.java#L391) that prevents any notifications after closed. It would be good not to rely on that, though, because it means somewhere else our code was in an inconsistent state, but in general it can be a good safety net. I see some bugs with it now though (like [we don't close the message if closed](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/ChannelImpl.java#L424)).

ejona86 commented:
Aaand now I am pretty certain `finishStream` is missing holding of the lock, at least when transportReportStatus is called. That should be fixed, but probably not in this PR, although it could.

ejona86 commented:
Hmm... In OkHttp, we really want the second argument to `transportReportStatus()` to be `true` when a deadline occurs, as that will cause it to clear any queued messages that have not yet been delivered to the application. It looks like we can [add an `|| status.getCode() == Status.DEADLINE_EXCEEDED`](https://github.com/grpc/grpc-java/blob/master/okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java#L462). Not great, but the alternatives look more painful.

For Netty you are already passing `true`.

ejona86 commented:
@johnbcoughlin, concerning the 120ms for first RPC, how did you get that number? I do see some numbers that high, but for the very first Channel on a cold JVM with TLS. I see better speeds with the first RPC of the second Channel (which would still be a mostly-cold JVM). Better speeds with plaintext.

johnbcoughlin commented:
@ejona86 the 120ms is definitely from the first channel on a cold JVM, in the `AbstractTransportTest`, which uses TLS. So yeah, perhaps that's fine.

Re: the contract to `closed()`, I think I understand what's going on. the `ClientStreamListenerImpl` inside the `ChannelImpl` is actually the one holding this non-volatile state, and it uses it to bail out of all of the callbacks early, if it's already been closed. That's a pretty neat trick with the `SerializingExecutor`, I like it.

In any case, it still doesn't really seem to me like a contract on the `ClientStreamListener` interface. Instead it's a contract that this particular instance of that interface makes with a `Listener<RespT>` with which it composes (the `observer` field). No callbacks to the `observer` will be issued after the `closed()` call. We might end up calling `closed()` twice but the second time would be a no-op, and no other callbacks will be interleaved. I think.

I'll clean up these tests tonight, but if I understand the concurrency questions correctly, this PR is almost ready for review already.

johnbcoughlin commented:
@ejona86 please take a look :)

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -386,6 +390,15 @@ private void completeHeaders(Metadata.Headers headers) {
     }
 
+    private void startDeadlineTimer() {
+      SharedResourceHolder.get(TIMER_SERVICE).schedule(new Runnable() {
```
Each get() should be coupled with a release(), as it does resource counting. I'd suggest doing a get() during construction and doing a release() near where terminationRunnable.run() is called (it is called twice).

ejona86 replied:
We should probably hang onto the returned `ScheduledFuture` and cancel it `ClientStreamListenerImpl.closed` when we set `closed = true`. That way we won't retain many old streams in memory if they had a long deadline. It looks like it will need to be a volatile reference, though.

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java:
```diff
@@ -586,6 +597,87 @@ public void sendsTimeoutHeader() {
   }
 
+  @Test
+  public void deadlineNotExceeded() {
```
I think ideally these would be in ChannelImplTest since this behavior doesn't really change based on underlying transport, but they are harder to write there and they are already here. So I don't think you need to move them, but just be aware for next time.

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java:
```diff
@@ -586,6 +597,87 @@ public void sendsTimeoutHeader() {
+  }
+
+  @Test(timeout = 10000)
+  public void deadlineAppliedAfterClientInterceptors() throws Exception {
```
This test we can probably get rid of. In #574 we are swapping to using a deadline, so this test would no longer make sense. #574 is pretty far along, and I expect will be merged today or tomorrow (and yes, it will cause some conflicts with this one, or vise versa).

ejona86 commented:
@johnbcoughlin Looks great! Some comments to be resolved, but it looks pretty close.

ejona86 commented:
I created #606 for improving the documentation on close() guarantees.

johnbcoughlin commented:
@ejona86 thanks for the review! Addressed your comments. I'm getting a checkstyle failure because the `ChannelImpl` constructor lacks javadoc, which is unexpected.

ejona86 commented:
The checkstyle failure is because public methods need JavaDoc. It probably didn't need it before because the method was deemed "trivial" by checkstyle.

You can add JavaDoc, or even better, just make it package-private which would fix #568 :-)

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -81,6 +85,7 @@
 
   private final ClientTransportFactory transportFactory;
   private final ExecutorService executor;
+  private volatile ScheduledExecutorService deadlineCancellationExecutor;
```
It looks like this could be non-`volatile`, and possibly even `final`.

ejona86 replied:
I see now it can't be `final`, but it doesn't need to be volatile as it would still only be accessed during the constructor or with the intrinsic lock held.

johnbcoughlin replied:
nice

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -119,6 +125,10 @@ public synchronized ChannelImpl shutdown() {
       return this;
     }
     shutdown = true;
+    if (deadlineCancellationExecutor != null) {
```
No need to check for `null`, since you never set it to `null`.

ejona86 replied:
Oh, nm. I see that you do set it to null.

johnbcoughlin replied:
That's the recommended pattern from `SharedResourceHolder.release`:

```
   * <p>Caller must not release a reference more than once. It's advisory that you clear the
   * reference to the instance with the null returned by this method.
```

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -119,6 +125,10 @@ public synchronized ChannelImpl shutdown() {
     }
     shutdown = true;
+    if (deadlineCancellationExecutor != null) {
+      deadlineCancellationExecutor =
```
This should probably happen when `terminated = true` below. That will help guarantee it only happens once (since elsewhere release code you added will run when `shutdown == true` but not when `terminated == true`)

ejona86 replied:
Actually, it _must_ be shutdown after termination, as RPCs are still functioning after `shutdown()` is called (just no new ones are permitted). So we will need deadline support even when shutdown. When `terminated` it is guaranteed there are no RPCs in progress, so it is safe to release this scheduler.

johnbcoughlin replied:
ah, thank you for the clarification, I should have investigated those guarantees more closely. :)

johnbcoughlin replied:
I think it might be okay to release at `shutdown`, actually. The executor will only be `shutdown` once the refcount reaches 0, not `shutdownNow()`, which will only reject new cancellation tasks, not interrupt those currently running. We only start a cancellation task when we start an RPC, so it should be safe to have a released executor after `shutdown = true`.

johnbcoughlin replied:
to check my work a little more: if `shutdown`, we'll never obtain an active transport in `start`, and therefore will enter `closeCallPrematurely`, rather than attempting to begin a new cancellation task.

ejona86 replied:
Ah, because `ExecuteExistingDelayedTasksAfterShutdownPolicy` defaults to `true`.

Yes, if `shutdown`, we will call `closeCallPrematurely()` because of `obtainActiveTransport()`'s return value.

ejona86 commented:
Making ChannelImpl's constructor package-private has already happened, as part of 8c1d38a.

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -281,6 +291,7 @@ public void transportTerminated() {
         transportShutdown();
         transports.remove(transport);
         if (shutdown && transports.isEmpty()) {
+          releaseDeadlineCancellationExecutor();
```
This is no longer needed. There is only one place when `shutdown` is set to `true`, and you are already covering it.

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -561,4 +590,34 @@ public Long parseAsciiString(String serialized) {
+   * Should only be called from methods that are themselves synchronized.
+   */
+  private synchronized void releaseDeadlineCancellationExecutor() {
+    if (deadlineCancellationExecutor != null) {
```
After removing the call to this from transportTerminated, the `!= null` would no longer be necessary and the method could easy be inlined in `shutdown()`.

ejona86 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -468,6 +494,9 @@ public void closed(final Status status, final Metadata.Trailers trailers) {
           @Override
           public void run() {
             closed = true;
+            if (deadlineCancellationFuture != null) {
```
`volatile` reads aren't optimized, so ideally you would read `deadlineCancellationFuture` into a local variable. It's fine as-is, but it is also good practice to use the local variable.

johnbcoughlin replied:
done, thanks for the tip!

ejona86 commented on lib/netty:
```diff
@@ -0,0 +1 @@
+Subproject commit 871ce43b1f431536e08a36d420875494ef3e9451
```
Remove this; we got rid of the submodule and this is residue left over in your checkout.

ejona86 commented:
@johnbcoughlin LGTM! I made some comments. Once you feel good about the PR state I'll merge it.

Thanks! The code looks great!

johnbcoughlin commented:
@ejona86 thanks for the review!

johnbcoughlin commented:
@ejona86 feel free to merge whenever is convenient!

ejona86 commented:
Cherry-picked as 3e26b99. Thanks!
